### PR TITLE
feat: support behavior training openpi model

### PR DESCRIPTION
**Pull request by XuS1994**
support behavior training openpi model

### Description

Rebase main branch. Add behavior policy and dataconfig.

New openpi is not finish yet because long time work can not get positive reward.

Rebuild behavior docker in RLinf dockehub.

### How has this been tested?

Test behavior+openpi evaluation.
Test behavior+openpi+grpo training.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/models/embodiment/openpi_action_model.py [hunk 1]
```diff
@@ -578,6 +582,53 @@ def get_suffix_out(
             "eager"  # noqa: SLF001
         )
 
+        # Fix attention mask size mismatch when using past_key_values
+        # When past_key_values is not None, key_states will include cached keys,
+        # so we need to extend attention_mask to match the actual key_states length
+        if past_key_values is not None:
```
Comment (LiuZhihao2022) [inline rlinf/models/embodiment/openpi_action_model.py]: Why should we process when past_key_values is not None

Comment (XuS1994) [reply rlinf/models/embodiment/openpi_action_model.py]: Handle `past_key_values` (cache) =>> assume `input_ids` just has unprocessed tokens

File: examples/embodiment/config/behavior_openpi_eval.yaml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  pipeline_stage_num: 1
+
+actor:
+  group_name: "ActorGroup"
+  channel:
+    name: ${env.channel.name}
+    queue_name: "replay_buffer"
+    queue_size: 0
+  training_backend: "fsdp"
+  checkpoint_load_path: "/path/to/model/RLinf-Pi0-SFTi"
```
Comment (QuanluZhang) [inline examples/embodiment/config/behavior_openpi_eval.yaml]: typo

Comment (XuS1994) [reply examples/embodiment/config/behavior_openpi_eval.yaml]: fixed

File: examples/embodiment/config/behavior_ppo_openpi.yaml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    log_path: "../results"
+    project_name: rlinf
+    experiment_name: "behavior_ppo_openpi"
+    logger_backends: ["tensorboard"] # wandb, swanlab
+
+  max_epochs: 1000
+  max_steps: -1
+
+  only_eval: False
+  eval_policy_path: null # Optional: .pt file or None, if None, will use the checkpoint in rollout.model.model_path
```
Comment (andylin-hao) [inline examples/embodiment/config/behavior_ppo_openpi.yaml]: I think this should align with this commit in main https://github.com/RLinf/RLinf/commit/acfee0441d31ca95212be4582f7723af8a4b4417#diff-b7194fa2e8b36d057d4996e11058353cca58ace93a8fde7fa5bb093292f6bf63L31

**Review by andylin-hao — APPROVED** [review]

File: examples/embodiment/config/behavior_ppo_openpi.yaml (hunk 3 above)
Comment (XuS1994) [reply examples/embodiment/config/behavior_ppo_openpi.yaml]: Fixed.

File: rlinf/envs/behavior/behavior_env.py [hunk 4]
```diff
@@ -125,12 +126,18 @@ def _extract_obs_image(self, raw_obs):
                     right_image = v["rgb"].to(torch.uint8)[..., :3]
                 elif "zed_link:Camera:0" in k:
                     zed_image = v["rgb"].to(torch.uint8)[..., :3]
+                elif "proprio" in k:
+                    state = v
+        assert state is not None, (
+            "state is not found in the observation which is required for the behavior training."
+        )
```
Comment (QuanluZhang) [inline rlinf/envs/behavior/behavior_env.py:133]: oft training does not require state?

File: rlinf/models/embodiment/openpi/dataconfig/__init__.py [hunk 5]
```diff
@@ -223,6 +226,21 @@
...
+        name="pi0_behavior",
+        model=pi0_config.Pi0Config(),
+        data=LeRobotBehaviorDataConfig(
+            repo_id="physical-intelligence/behavior",
+            base_config=DataConfig(prompt_from_task=True),
+            assets=AssetsConfig(assets_dir="checkpoints/torch/pi0_behavior/assets"),
+            extra_delta_transform=True,
+        ),
+        weight_loader=weight_loaders.CheckpointWeightLoader(
+            "checkpoints/jax/pi0_base/params"
```
Comment (QuanluZhang) [inline rlinf/models/embodiment/openpi/dataconfig/__init__.py:239]: do I need to provide jax format weights?

File: tests/e2e_tests/embodied/behavior_ppo_openpi.yaml [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+actor:
+  group_name: "ActorGroup"
+  training_backend: "fsdp"
+  micro_batch_size: 2
+  global_batch_size: 16
+  seed: 42
+  enable_offload: False
+
+  model:
+    model_path: "/workspace/dataset/RLinf-Pi0-BEHAVIOR"
```
Comment (QuanluZhang) [inline tests/e2e_tests/embodied/behavior_ppo_openpi.yaml:128]: do we have to use our transformed checkpoint? how can I use the model weights from openpi-comet?

File: rlinf/envs/behavior/behavior_env.py (hunk 4 above)
Comment (XuS1994) [reply rlinf/envs/behavior/behavior_env.py:133]: yes, 'states' is supported by behavior env, only 'opnepi' model needs it as input for training.

File: rlinf/models/embodiment/openpi/dataconfig/__init__.py (hunk 5 above)
Comment (XuS1994) [reply rlinf/models/embodiment/openpi/dataconfig/__init__.py:239]: No, it can only load torch format model.

We can not load jax format model. We need to convert jax to pytorch model by such file [toolkits/ckpt_convertor/convert_openpi_jax_to_python.py](https://github.com/RLinf/RLinf/pull/258/files#diff-c0d6ecb69f09b548d404284b35633cbfafbddc0c37bcdf5fa3686d009605a496)

File: tests/e2e_tests/embodied/behavior_ppo_openpi.yaml (hunk 6 above)
Comment (XuS1994) [reply tests/e2e_tests/embodied/behavior_ppo_openpi.yaml:128]: An error occurs when we can directly load the opnepi torch model, indicating that "the norm_state.json file cannot be found."

Using our format can solve this problem because `norm_state.json` saved in `RLinf-Pi0-BEHAVIOR/physical-intelligence/behavior`
